File: frontend/types/app-schema.ts
/**
 * MechaStream — Schema-first app definition.
 * AI may only output components from the CONTROLLED_REGISTRY.
 * Used for: Prompt → JSON Schema → Deterministic Code Builder
 */

/** Allowed component names only. AI cannot invent components outside this list. */
export const CONTROLLED_COMPONENT_NAMES = [
  'Hero',
  'Navbar',
  'Features',
  'Pricing',
  'Testimonials',
  'CTA',
  'Footer',
  'Dashboard',
  'Table',
  'Form',
  'Card',
  'Modal',
  'Sidebar',
  'Stats',
  'Chart',
] as const;

export type ComponentName = (typeof CONTROLLED_COMPONENT_NAMES)[number];

export type AppType = 'landing' | 'dashboard' | 'portfolio' | 'saas';
export type BorderRadius = 'sharp' | 'rounded' | 'pill';
export type Spacing = 'compact' | 'normal' | 'relaxed';
export type FontFamily = 'inter' | 'poppins' | 'roboto' | 'manrope';

export interface ThemeSchema {
  primaryColor: string;
  fontFamily: FontFamily | string;
  borderRadius: BorderRadius;
  spacing: Spacing;
}

export interface MetaSchema {
  title: string;
  type: AppType;
  theme: ThemeSchema;
}

export interface SectionSchema {
  component: ComponentName;
  variant: string;
  props: Record<string, unknown>;
}

export interface PageSchema {
  name: string;
  route: string;
  sections: SectionSchema[];
}

export interface AppSchema {
  meta: MetaSchema;
  pages: PageSchema[];
}

/** Type guard: check if a string is a valid ComponentName */
export function isComponentName(value: string): value is ComponentName {
  return (CONTROLLED_COMPONENT_NAMES as readonly string[]).includes(value);
}

/** Validate minimal AppSchema shape (for runtime checks before Pydantic). */
export function isAppSchemaLike(value: unknown): value is AppSchema {
  if (!value || typeof value !== 'object') return false;
  const o = value as Record<string, unknown>;
  if (!o.meta || typeof o.meta !== 'object') return false;
  const meta = o.meta as Record<string, unknown>;
  if (typeof meta.title !== 'string' || typeof meta.type !== 'string') return false;
  if (!o.pages || !Array.isArray(o.pages)) return false;
  for (const page of o.pages as unknown[]) {
    if (!page || typeof page !== 'object') return false;
    const p = page as Record<string, unknown>;
    if (typeof p.name !== 'string' || typeof p.route !== 'string' || !Array.isArray(p.sections))
      return false;
    for (const sec of (p.sections as unknown[]) || []) {
      if (!sec || typeof sec !== 'object') return false;
      const s = sec as Record<string, unknown>;
      if (!isComponentName(String(s.component)) || typeof s.variant !== 'string') return false;
    }
  }
  return true;
}

File: backend/app_schema.py
"""
MechaStream — Schema-first app definition (Pydantic v1).
Must stay in sync with frontend src/types/app-schema.ts.
Used for: Prompt → JSON Schema → Deterministic Code Builder.
"""

from enum import Enum
from typing import Any, Dict, List

from pydantic import BaseModel, Field, validator


# ─── Controlled component registry (AI cannot invent components outside this list) ───
CONTROLLED_COMPONENT_NAMES = frozenset({
    "Hero", "Navbar", "Features", "Pricing", "Testimonials",
    "CTA", "Footer", "Dashboard", "Table", "Form",
    "Card", "Modal", "Sidebar", "Stats", "Chart",
})


class AppType(str, Enum):
    landing = "landing"
    dashboard = "dashboard"
    portfolio = "portfolio"
    saas = "saas"


class BorderRadius(str, Enum):
    sharp = "sharp"
    rounded = "rounded"
    pill = "pill"


class Spacing(str, Enum):
    compact = "compact"
    normal = "normal"
    relaxed = "relaxed"


class ThemeSchema(BaseModel):
    primaryColor: str = Field(..., min_length=1, max_length=64)
    fontFamily: str = Field(..., min_length=1, max_length=128)
    borderRadius: BorderRadius = BorderRadius.rounded
    spacing: Spacing = Spacing.normal


class MetaSchema(BaseModel):
    title: str = Field(..., min_length=1, max_length=120)
    type: AppType = AppType.landing
    theme: ThemeSchema


class SectionSchema(BaseModel):
    component: str = Field(..., min_length=1, max_length=32)
    variant: str = Field(default="default", max_length=64)
    props: Dict[str, Any] = Field(default_factory=dict)

    @validator("component")
    def component_must_be_in_registry(cls, v: str) -> str:
        if v not in CONTROLLED_COMPONENT_NAMES:
            allowed = ", ".join(sorted(CONTROLLED_COMPONENT_NAMES))
            raise ValueError(
                f"component '{v}' is not in the controlled registry. "
                f"Allowed: {allowed}"
            )
        return v


class PageSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=80)
    route: str = Field(..., min_length=1, max_length=128)
    sections: List[SectionSchema] = Field(default_factory=list)


class AppSchema(BaseModel):
    meta: MetaSchema
    pages: List[PageSchema] = Field(...)

    @validator("pages")
    def at_least_one_page(cls, v: List[PageSchema]) -> List[PageSchema]:
        if not v:
            raise ValueError("At least one page is required")
        return v


def validate_app_schema(data: Dict[str, Any] | str) -> AppSchema:
    """
    Parse and validate JSON against AppSchema.
    Raises pydantic.ValidationError if invalid.
    """
    if isinstance(data, str):
        import json
        data = json.loads(data)
    return AppSchema.parse_obj(data)
